client closed request should not be reported as 5xx

szuecs opened the pull request:
fix #687: client closed request should not be reported as 5xx

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

aermakov-zalando commented:
Deadline exceeded sounds more like 502 gateway timeout, no? Only client cancels should be reported as 499.

aryszka commented on proxy/proxy.go:
```diff
@@ -408,10 +408,11 @@ func (dc *skipperDialer) DialContext(ctx stdlibcontext.Context, network, addr st
 		return nil, &proxyError{
 			err:  cerr,
-			code: http.StatusGatewayTimeout,
+			code: 499,
```
this is not a standard code. Do we have other options?

aermakov-zalando replied:
This is used by nginx (and I think apache, but not sure).

aryszka replied:
maybe 0? indicating that the request was not served, therefore there is no status code at all?

szuecs replied:
0 is specially handled in skipper and returns in the end 500 @aryszka

mikkeloscar commented:
I don't see a 499 logged when running skipper build from this branch. I do as described in https://github.com/zalando/skipper/issues/687#issuecomment-405557503 (Thanks for a very good description by the way).

I only see 499 if I replace the 504 here: https://github.com/zalando/skipper/blob/e9c64ec7e2a8de4ad85f7163807683c446ba80f5/proxy/proxy.go#L730

szuecs commented:
@aermakov-zalando maybe my message was missleading, but it's exactly client cacneled the request, see https://github.com/zalando/skipper/issues/687#issuecomment-405557503

@mikkeloscar gotcha, thanks for testing!

szuecs commented:
I fixed the issue reported by @mikkeloscar.
I additionally cleaned up the logging and enhanced the error message: dropped duplicates and added the backend to the log, which was an addition from the middle log entry.

proxy:
```
 % ./bin/skipper --address :9999 -inline-routes 'p0: * -> "http://127.0.0.1:9090/"'
[APP]INFO[0000] Expose metrics in codahale format
[APP]INFO[0000] support listener on :9911
[APP]INFO[0000] proxy listener on :9999
[APP]INFO[0000] certPathTLS or keyPathTLS not found, defaulting to HTTP
[APP]INFO[0000] route settings, reset, route: p0: * -> "http://127.0.0.1:9090/"
[APP]INFO[0000] route settings received
[APP]INFO[0000] route settings applied
[APP]ERRO[0007] error while proxying, route p0 with backend http://127.0.0.1:9090/, status code 499: dialing failed false: context canceled
::1 - - [17/Jul/2018:18:28:36 +0200] "GET / HTTP/1.1" 499 1 "-" "curl/7.49.0" 1000 localhost:9999 - -
::1 - - [17/Jul/2018:18:28:35 +0200] "GET / HTTP/1.1" 200 14 "-" "curl/7.49.0" 5001 localhost:9999 - -
```

backend:
```
% ./bin/skipper -inline-routes 'p0: * -> latency(5000) -> inlineContent("<h1>Hello</h1>") -> <shunt>'
[APP]INFO[0000] Expose metrics in codahale format
[APP]INFO[0000] route settings, reset, route: p0: * -> latency(5000) -> inlineContent("<h1>Hello</h1>") -> <shunt>
[APP]INFO[0000] route settings received
[APP]INFO[0000] route settings applied
[APP]INFO[0000] support listener on :9911
[APP]INFO[0000] proxy listener on :9090
[APP]INFO[0000] certPathTLS or keyPathTLS not found, defaulting to HTTP


127.0.0.1 - - [17/Jul/2018:18:28:35 +0200] "GET / HTTP/1.1" 200 14 "-" "curl/7.49.0" 5000 127.0.0.1:9090 - -
127.0.0.1 - - [17/Jul/2018:18:28:36 +0200] "GET / HTTP/1.1" 200 14 "-" "curl/7.49.0" 5000 127.0.0.1:9090 - -
```

client calls:
```
% curl localhost:9999/ &
[1] 29258
[sszuecs@sandor-lab:~]% curl -m 1 localhost:9999/
curl: (28) Operation timed out after 1000 milliseconds with 0 bytes received
zsh: exit 28    curl -m 1 localhost:9999/
[sszuecs@sandor-lab:~]% <h1>Hello</h1>
[1]  + done       curl localhost:9999/
```

mikkeloscar commented on proxy/proxy.go:
```diff
@@ -408,10 +409,11 @@ func (dc *skipperDialer) DialContext(ctx stdlibcontext.Context, network, addr st
 		return nil, &proxyError{
-			err:  cerr,
-			code: http.StatusGatewayTimeout,
+			err:  fmt.Errorf("err from dial contrext: %v", cerr),
```
`contrext` -> `context`.

mikkeloscar commented on proxy/proxy.go:
```diff
@@ -408,10 +409,11 @@ func (dc *skipperDialer) DialContext(ctx stdlibcontext.Context, network, addr st
-			err:  cerr,
-			code: http.StatusGatewayTimeout,
+			err:  fmt.Errorf("err from dial contrext: %v", cerr),
+			code: 499,
```
Are we sure this is always caused by client closing the connection?

I'm just afraid that we are masking some `http: server closed idle connection` errors as mentioned in https://github.com/zalando/skipper/issues/687 but I'm also not sure where they came from.

szuecs commented:
fixed comments and squashed commits.

The only open question is if 499 is fine or someone disagrees. 

My reasoning:
The error is raised, because the client has a timeout. The proxy did not timedout and get's the value from the backend and will respond to the client which already send a packet with a FIN flag (socket half closed), which will result to the client will respond the proxy response with a TCP RST packet, because it already closed the socket from the client and the kernel will send a RST packet. From the proxy point of view everything worked fine, but the client hang up. For me this is a 4xx and since nginx uses 499 I would propose to reuse their http status code, because having another value does not makes sense to me and probably some folks are already familiar of the meaning of 499.

aryszka commented on proxy/proxy.go:
```diff
@@ -723,11 +724,11 @@ func (p *Proxy) makeBackendRequest(ctx *context) (*http.Response, *proxyError) {
-			return nil, &proxyError{err: cerr, code: http.StatusGatewayTimeout}
+			// deadline exceeded or canceled in stdlib, client closed request
+			// see https://github.com/zalando/skipper/issues/687#issuecomment-405557503
+			return nil, &proxyError{err: cerr, code: 499}
```
if this is a request that the client has already cancelled, then this status code is used only for logging. Is this correct? If yes, then i'd prefer leaving the status code 0, because it is something that we use only for logging.

mikkeloscar replied:
It's also added to the metrics, so that's why a 5xx is a problem because it gives the wrong impression of the service error rate.

Counting this as a 4xx seems right because we can exclude it from errors caused by the proxy/backend, but we can still monitor the 4xx rate and be aware of the error rate.

It would be interesting to know what AWS ELBs report in this case.

szuecs replied:
Do you think the AWS ELB is better for the status code handling then nginx?
I mean AWS ELB also have a very different way of handling X-For-* headers and set the clientip last instead of all others that put this first as shown in an example of the RFC.

mikkeloscar replied:
> Do you think the AWS ELB is better for the status code handling then nginx?

AWS ELB is just interesting because this is what our users are used to.

> For me the biggest drawback to status code 0 is that:

Note that status code `0` will be reported as `500` in logs and metrics. This is my biggest concern because it will give the wrong impression to users.

I agree with the benefit of `499` being fairly simple to search for.

aryszka replied:
i got convinced, thanks for explaining it. Let's use 499.

aryszka commented:
:+1:

mikkeloscar commented:
:+1:
